## Импорт библиотек

In [373]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

## Загрузка датасетов

In [374]:
df_chat = pd.read_csv('Data/chat_messages.csv')
df_chat.head()

message_id                   type  entity_id  created_by  \
0  "01jb7da570sf4f65xdf0ptvv9q"  incoming_chat_message   37556493           0   
1  "01jb3t6bc8gvyfnd063nd91ng4"  incoming_chat_message   37549491           0   
2  "01jb6gj9ngwk0ybbmg9w90pbqg"  incoming_chat_message   37531455           0   
3  "01jb75bkprk0d7hht1g5vy1pm4"  incoming_chat_message   37553371           0   
4  "01jb3t5xprbph1c6pym8ervxpm"  incoming_chat_message   37548675           0   

   created_at  
0  1730046924  
1  1729926213  
2  1730016782  
3  1730038583  
4  1729926199

In [375]:
df_managers = pd.read_csv('Data/managers.csv')
df_managers

mop_id             name_mop  rop_id
0    6645315       Гюнель и Илина       1
1    6744792         Юля и Наташа       1
2   10262505         Вика и Марго       1
3   10262513           Ира и Варя       1
4   10465254         Настя и Даша       1
5   11448610          Вика и Катя       1
6   11490270          Аня и Ксюша       1
7   11490278      Ксюша и Джамиля       1
8    6780177          Даша и Даша       2
9    7436430         Софа и Алина       2
10   7996770           Лиза и Ева       2
11  10262485          Соня и Катя       2
12  10262497         Мария и Соня       2
13  11188802         Даша и Влада       2
14  11394130         Настя и Саша       2
15   7408305        Влада и Настя       3
16   7408332             Пустой 2       3
17   7417617     Порхачева Полина       3
18   8558094          Алина и Юля       3
19  10262493          Ками и Мила       3
20  10465690       Настя и Малика       3
21  11241942        Даша и Карина       3
22  11394126       Полина Мирзоян       3
23  11394134         Софья Боднар       3
24  11503694     Ахмедова Джамиля       3
25  11550890  Ангелина Милованова       3

In [376]:
df_rops = pd.read_csv('Data/rops.csv')
df_rops

rop_id    rop_name
0       1    Катя РОП
1       2  Полина РОП
2       3     Эля РОП

## Выполнение задания

Будем работать с копией датасета df_chat

In [377]:
df = df_chat.copy()

In [378]:
df[(df['created_by'] == 0) & (df['type'] == 'outgoing_chat_message')]

message_id                   type  entity_id  \
143    "01jb3sy7sddzcrvsd63sdr0ed0"  outgoing_chat_message   37188715   
581    "01jb3v4e9y6a9rzemygkf1aeyz"  outgoing_chat_message   37549551   
733    "01jb3v417ch32w53ahpd58xcww"  outgoing_chat_message   37549551   
760    "01jb3veh4p86seakzf10zcrxjt"  outgoing_chat_message   37549569   
761    "01jb3vf5636d3asf65dbn6m6g9"  outgoing_chat_message   37549569   
...                             ...                    ...        ...   
18681  "01jb7pqkxdkkbjhv7epkzcv6gm"  outgoing_chat_message   37557047   
18726  "01jb7ste5bmf21nx491cc4xfk0"  outgoing_chat_message   36541651   
18750  "01jb7smdg6ede3hf4epma9bsf0"  outgoing_chat_message   36541651   
18760  "01jb7t0yknb9dwd08dbcb8abvf"  outgoing_chat_message   36541673   
18858  "01jb7vehzh14a70ttx6t7z1dst"  outgoing_chat_message   37548639   

       created_by  created_at  
143             0  1729925947  
581             0  1729927199  
733             0  1729927185  
760             0  1729927529  
761             0  1729927550  
...           ...         ...  
18681           0  1730056802  
18726           0  1730060040  
18750           0  1730059843  
18760           0  1730060253  
18858           0  1730061748  

[173 rows x 5 columns]

У нас выявлено 173 сообщения, которые помечены, как отправленным менеджером, но с идентификатором 0. Это явный сбой в системе. 
Так как в условии ясно сказано, что идентификатор 0 всегда у клиента, я решил в этих 173 строках изменить тип на входящий, отправленный клиентом менеджеру

In [379]:
df.type.value_counts()

type
incoming_chat_message    10140
outgoing_chat_message     8789
Name: count, dtype: int64

In [380]:
df['type'] = df.apply(lambda x: 'incoming_chat_message' if (x.created_by == 0) & (x.type == 'outgoing_chat_message') else x.type, axis=1)
df.type.value_counts()

type
incoming_chat_message    10313
outgoing_chat_message     8616
Name: count, dtype: int64

Далее переведем время сообщения из формата UnixTimestamp в тип Даты и времени с учетом часового пояса

In [381]:
df.loc[:, ['created_at']] = df['created_at'].apply(lambda x: pd.Timestamp(x, unit='s', tz='Europe/Moscow'))
df.head()

message_id                   type  entity_id  created_by  \
0  "01jb7da570sf4f65xdf0ptvv9q"  incoming_chat_message   37556493           0   
1  "01jb3t6bc8gvyfnd063nd91ng4"  incoming_chat_message   37549491           0   
2  "01jb6gj9ngwk0ybbmg9w90pbqg"  incoming_chat_message   37531455           0   
3  "01jb75bkprk0d7hht1g5vy1pm4"  incoming_chat_message   37553371           0   
4  "01jb3t5xprbph1c6pym8ervxpm"  incoming_chat_message   37548675           0   

                 created_at  
0 2024-10-27 19:35:24+03:00  
1 2024-10-26 10:03:33+03:00  
2 2024-10-27 11:13:02+03:00  
3 2024-10-27 17:16:23+03:00  
4 2024-10-26 10:03:19+03:00

Затем отсортируемся по номеру сделки и времени для лучшего понимания

In [382]:
df = df.sort_values(by=['entity_id', 'created_at'])
df.head(20)

message_id                   type  entity_id  \
3107   "01jb4pjb1g22ts7z6arcjv11w6"  incoming_chat_message   35810509   
4102   "01jb4pncn7gm1re1zdxkzhnp6s"  outgoing_chat_message   35810509   
6430   "01jb4qedfgxnd2yqr4w26hs4q5"  incoming_chat_message   35810509   
6639   "01jb4qj2skc2wjfdm8qaf69yjh"  outgoing_chat_message   35810509   
17586  "01jb7gxwb0sepv9vtekrgj3v6d"  incoming_chat_message   35945535   
17852  "01jb7h4ad8weqjx1ynrh7s8k25"  outgoing_chat_message   35945535   
14805  "01jb727x90ga4ncv4m705rh0rm"  incoming_chat_message   36099337   
14833  "01jb72ezmaacke29h682geasez"  outgoing_chat_message   36099337   
14989  "01jb72hktgm0tvdpwk4a17dadz"  incoming_chat_message   36099337   
15039  "01jb72nzkrzhcfhg9hfpbjt0vx"  outgoing_chat_message   36099337   
16568  "01jb789nf89v0wchbsh9cnm4n5"  incoming_chat_message   36209333   
429    "01jb4b9g5g3j9rjkx8c41w55vg"  incoming_chat_message   36222379   
6282   "01jb4ft2b502923xpqhn00ss7r"  outgoing_chat_message   36308085   
9089   "01jb4th8j8zztbcz0c68xq5pkf"  incoming_chat_message   36333761   
9238   "01jb4tsxdn395tp1jwknjajpwk"  outgoing_chat_message   36333761   
9306   "01jb4v0dxgwsz4cqxg6wted3s2"  incoming_chat_message   36333761   
9569   "01jb4vbfpfwe4my2w5zhadeash"  outgoing_chat_message   36333761   
8311   "01jb4vc5wrcvrpxggjendk442q"  incoming_chat_message   36333761   
9547   "01jb4ve204kyab9ez1hb5d2e1f"  outgoing_chat_message   36333761   
9730   "01jb4vffbryvgpbke26kzya92a"  incoming_chat_message   36333761   

       created_by                created_at  
3107            0 2024-10-26 18:19:26+03:00  
4102     10262493 2024-10-26 18:21:05+03:00  
6430            0 2024-10-26 18:34:46+03:00  
6639     10262493 2024-10-26 18:36:46+03:00  
17586           0 2024-10-27 20:38:36+03:00  
17852    10262493 2024-10-27 20:42:07+03:00  
14805           0 2024-10-27 16:21:56+03:00  
14833    11241942 2024-10-27 16:25:47+03:00  
14989           0 2024-10-27 16:27:14+03:00  
15039    11241942 2024-10-27 16:29:37+03:00  
16568           0 2024-10-27 18:07:45+03:00  
429             0 2024-10-26 15:02:22+03:00  
6282      6744792 2024-10-26 16:21:19+03:00  
9089            0 2024-10-26 19:28:45+03:00  
9238     10262485 2024-10-26 19:33:28+03:00  
9306            0 2024-10-26 19:37:02+03:00  
9569     10262485 2024-10-26 19:43:04+03:00  
8311            0 2024-10-26 19:43:27+03:00  
9547     10262485 2024-10-26 19:44:28+03:00  
9730            0 2024-10-26 19:45:15+03:00

Для следующих действий необходимо получить дополнительные столбцы: Время и тип сообщения на предыдущей строке.

In [383]:
df['previous_type'] = df.groupby('entity_id')['type'].shift()
df['previous_time'] = df.groupby('entity_id')['created_at'].shift()
df.head(20)

message_id                   type  entity_id  \
3107   "01jb4pjb1g22ts7z6arcjv11w6"  incoming_chat_message   35810509   
4102   "01jb4pncn7gm1re1zdxkzhnp6s"  outgoing_chat_message   35810509   
6430   "01jb4qedfgxnd2yqr4w26hs4q5"  incoming_chat_message   35810509   
6639   "01jb4qj2skc2wjfdm8qaf69yjh"  outgoing_chat_message   35810509   
17586  "01jb7gxwb0sepv9vtekrgj3v6d"  incoming_chat_message   35945535   
17852  "01jb7h4ad8weqjx1ynrh7s8k25"  outgoing_chat_message   35945535   
14805  "01jb727x90ga4ncv4m705rh0rm"  incoming_chat_message   36099337   
14833  "01jb72ezmaacke29h682geasez"  outgoing_chat_message   36099337   
14989  "01jb72hktgm0tvdpwk4a17dadz"  incoming_chat_message   36099337   
15039  "01jb72nzkrzhcfhg9hfpbjt0vx"  outgoing_chat_message   36099337   
16568  "01jb789nf89v0wchbsh9cnm4n5"  incoming_chat_message   36209333   
429    "01jb4b9g5g3j9rjkx8c41w55vg"  incoming_chat_message   36222379   
6282   "01jb4ft2b502923xpqhn00ss7r"  outgoing_chat_message   36308085   
9089   "01jb4th8j8zztbcz0c68xq5pkf"  incoming_chat_message   36333761   
9238   "01jb4tsxdn395tp1jwknjajpwk"  outgoing_chat_message   36333761   
9306   "01jb4v0dxgwsz4cqxg6wted3s2"  incoming_chat_message   36333761   
9569   "01jb4vbfpfwe4my2w5zhadeash"  outgoing_chat_message   36333761   
8311   "01jb4vc5wrcvrpxggjendk442q"  incoming_chat_message   36333761   
9547   "01jb4ve204kyab9ez1hb5d2e1f"  outgoing_chat_message   36333761   
9730   "01jb4vffbryvgpbke26kzya92a"  incoming_chat_message   36333761   

       created_by                created_at          previous_type  \
3107            0 2024-10-26 18:19:26+03:00                    NaN   
4102     10262493 2024-10-26 18:21:05+03:00  incoming_chat_message   
6430            0 2024-10-26 18:34:46+03:00  outgoing_chat_message   
6639     10262493 2024-10-26 18:36:46+03:00  incoming_chat_message   
17586           0 2024-10-27 20:38:36+03:00                    NaN   
17852    10262493 2024-10-27 20:42:07+03:00  incoming_chat_message   
14805           0 2024-10-27 16:21:56+03:00                    NaN   
14833    11241942 2024-10-27 16:25:47+03:00  incoming_chat_message   
14989           0 2024-10-27 16:27:14+03:00  outgoing_chat_message   
15039    11241942 2024-10-27 16:29:37+03:00  incoming_chat_message   
16568           0 2024-10-27 18:07:45+03:00                    NaN   
429             0 2024-10-26 15:02:22+03:00                    NaN   
6282      6744792 2024-10-26 16:21:19+03:00                    NaN   
9089            0 2024-10-26 19:28:45+03:00                    NaN   
9238     10262485 2024-10-26 19:33:28+03:00  incoming_chat_message   
9306            0 2024-10-26 19:37:02+03:00  outgoing_chat_message   
9569     10262485 2024-10-26 19:43:04+03:00  incoming_chat_message   
8311            0 2024-10-26 19:43:27+03:00  outgoing_chat_message   
9547     10262485 2024-10-26 19:44:28+03:00  incoming_chat_message   
9730            0 2024-10-26 19:45:15+03:00  outgoing_chat_message   

                  previous_time  
3107                        NaT  
4102  2024-10-26 18:19:26+03:00  
6430  2024-10-26 18:21:05+03:00  
6639  2024-10-26 18:34:46+03:00  
17586                       NaT  
17852 2024-10-27 20:38:36+03:00  
14805                       NaT  
14833 2024-10-27 16:21:56+03:00  
14989 2024-10-27 16:25:47+03:00  
15039 2024-10-27 16:27:14+03:00  
16568                       NaT  
429                         NaT  
6282                        NaT  
9089                        NaT  
9238  2024-10-26 19:28:45+03:00  
9306  2024-10-26 19:33:28+03:00  
9569  2024-10-26 19:37:02+03:00  
8311  2024-10-26 19:43:04+03:00  
9547  2024-10-26 19:43:27+03:00  
9730  2024-10-26 19:44:28+03:00

Теперь на основе созданных столбцов мы можем получить новый признак с времени исходящего сообщения для каждого ответа. При этом мы учитываем, чтобы значения создавались только для исходящий сообщений и тип сообщения не был одинаков на двух и более строках подряд. Это поможет нам учесть случаи, когда несколько подряд сообщений могут идти от одного адресата.
В конце удалим вспомогательные столбцы

In [384]:
df['question_time'] = df.apply(lambda x: x['previous_time'] if (x['type'] != x['previous_type']) & (x['type'] == 'outgoing_chat_message') else None, axis=1)
df = df.drop(['previous_type', 'previous_time'], axis=1)
df.head(20)

message_id                   type  entity_id  \
3107   "01jb4pjb1g22ts7z6arcjv11w6"  incoming_chat_message   35810509   
4102   "01jb4pncn7gm1re1zdxkzhnp6s"  outgoing_chat_message   35810509   
6430   "01jb4qedfgxnd2yqr4w26hs4q5"  incoming_chat_message   35810509   
6639   "01jb4qj2skc2wjfdm8qaf69yjh"  outgoing_chat_message   35810509   
17586  "01jb7gxwb0sepv9vtekrgj3v6d"  incoming_chat_message   35945535   
17852  "01jb7h4ad8weqjx1ynrh7s8k25"  outgoing_chat_message   35945535   
14805  "01jb727x90ga4ncv4m705rh0rm"  incoming_chat_message   36099337   
14833  "01jb72ezmaacke29h682geasez"  outgoing_chat_message   36099337   
14989  "01jb72hktgm0tvdpwk4a17dadz"  incoming_chat_message   36099337   
15039  "01jb72nzkrzhcfhg9hfpbjt0vx"  outgoing_chat_message   36099337   
16568  "01jb789nf89v0wchbsh9cnm4n5"  incoming_chat_message   36209333   
429    "01jb4b9g5g3j9rjkx8c41w55vg"  incoming_chat_message   36222379   
6282   "01jb4ft2b502923xpqhn00ss7r"  outgoing_chat_message   36308085   
9089   "01jb4th8j8zztbcz0c68xq5pkf"  incoming_chat_message   36333761   
9238   "01jb4tsxdn395tp1jwknjajpwk"  outgoing_chat_message   36333761   
9306   "01jb4v0dxgwsz4cqxg6wted3s2"  incoming_chat_message   36333761   
9569   "01jb4vbfpfwe4my2w5zhadeash"  outgoing_chat_message   36333761   
8311   "01jb4vc5wrcvrpxggjendk442q"  incoming_chat_message   36333761   
9547   "01jb4ve204kyab9ez1hb5d2e1f"  outgoing_chat_message   36333761   
9730   "01jb4vffbryvgpbke26kzya92a"  incoming_chat_message   36333761   

       created_by                created_at             question_time  
3107            0 2024-10-26 18:19:26+03:00                       NaT  
4102     10262493 2024-10-26 18:21:05+03:00 2024-10-26 18:19:26+03:00  
6430            0 2024-10-26 18:34:46+03:00                       NaT  
6639     10262493 2024-10-26 18:36:46+03:00 2024-10-26 18:34:46+03:00  
17586           0 2024-10-27 20:38:36+03:00                       NaT  
17852    10262493 2024-10-27 20:42:07+03:00 2024-10-27 20:38:36+03:00  
14805           0 2024-10-27 16:21:56+03:00                       NaT  
14833    11241942 2024-10-27 16:25:47+03:00 2024-10-27 16:21:56+03:00  
14989           0 2024-10-27 16:27:14+03:00                       NaT  
15039    11241942 2024-10-27 16:29:37+03:00 2024-10-27 16:27:14+03:00  
16568           0 2024-10-27 18:07:45+03:00                       NaT  
429             0 2024-10-26 15:02:22+03:00                       NaT  
6282      6744792 2024-10-26 16:21:19+03:00                       NaT  
9089            0 2024-10-26 19:28:45+03:00                       NaT  
9238     10262485 2024-10-26 19:33:28+03:00 2024-10-26 19:28:45+03:00  
9306            0 2024-10-26 19:37:02+03:00                       NaT  
9569     10262485 2024-10-26 19:43:04+03:00 2024-10-26 19:37:02+03:00  
8311            0 2024-10-26 19:43:27+03:00                       NaT  
9547     10262485 2024-10-26 19:44:28+03:00 2024-10-26 19:43:27+03:00  
9730            0 2024-10-26 19:45:15+03:00                       NaT

Далее я хотел продемонстрировать исходящие ответы менеджеров, которые были отправлены с полуночи до начала рабочего времени.

In [385]:
df[(df['created_at'].dt.time.between(pd.Timestamp('00:00:00').time(), pd.Timestamp('09:30:00').time())) & (df['type'] == 'outgoing_chat_message') & (df['question_time'].notna())]

message_id                   type  entity_id  \
10547  "01jb6adny7m9tz11tnbjtbmt7m"  outgoing_chat_message   37054485   
11512  "01jb5a6p5hya4wrsdsdgd6brwd"  outgoing_chat_message   37192445   
1538   "01jb3qctsjd1c849g6fx907vw7"  outgoing_chat_message   37209813   
11645  "01jb5dmr4ztmkrjns9ntyx4jgh"  outgoing_chat_message   37456077   
868    "01jb2qtqzx64r807w7s5xest28"  outgoing_chat_message   37471037   
11599  "01jb5a8ytt2vs0921ceeyg1hmv"  outgoing_chat_message   37485197   
10545  "01jb6afrfddc3xdkdghvm25t1j"  outgoing_chat_message   37494631   
1562   "01jb3r7bn88r3nv8yyfrr34hhr"  outgoing_chat_message   37497027   
11510  "01jb5a6vm8w8vkdzmfceeeks60"  outgoing_chat_message   37504219   
11514  "01jb5a5qrkqssmqgqk0jab5cy1"  outgoing_chat_message   37511647   
11596  "01jb5abr8ydrfqz4msdtkenx3j"  outgoing_chat_message   37515333   
10544  "01jb6agj3t977jhwahrqhmcz6c"  outgoing_chat_message   37515623   
11518  "01jb5a39agvaag2c0tfkjsqwp3"  outgoing_chat_message   37522043   
792    "01jb2qste3s1ffq030d0taxjym"  outgoing_chat_message   37529075   
867    "01jb2qvdmf61abj81ex9w092pr"  outgoing_chat_message   37529075   
1391   "01jb3mvvy8ps4wghqpk5e54f03"  outgoing_chat_message   37534819   
11598  "01jb5a9pzqncb8tjeyazgd78x1"  outgoing_chat_message   37535255   
8485   "01jb68rv3q18resfnz0bp161qz"  outgoing_chat_message   37535387   
1564   "01jb3r5jyssejgpqqts6n7a6zv"  outgoing_chat_message   37537651   
887    "01jb2r6gdg26zqpm2p85pfa7yr"  outgoing_chat_message   37546825   
793    "01jb2qrdx8khp0cwr5nh69zfct"  outgoing_chat_message   37548087   
1560   "01jb3r8pbz0tj8zgmyehrns4za"  outgoing_chat_message   37548761   
11950  "01jb673qvep0q8jamht5f1ykxj"  outgoing_chat_message   37549073   
7807   "01jb68e1f1tk7wc4dahf7rk6gn"  outgoing_chat_message   37550847   
7233   "01jb684wg7e8t3rdhk7dhgtwqy"  outgoing_chat_message   37551297   
11516  "01jb5a4fkk823g4tj7xm97rnk6"  outgoing_chat_message   37551471   
11513  "01jb5a691ns3w8j6tm9srqmr3t"  outgoing_chat_message   37551471   
10541  "01jb6ahr45pqx3117ks7y84wje"  outgoing_chat_message   37552419   
11595  "01jb5ae2eb9rekb2vqrqtd53p6"  outgoing_chat_message   37553097   
11597  "01jb5abj9g15pj9wpkx5kyxmx5"  outgoing_chat_message   37553371   
11521  "01jb5a332v9gn4wcgq6t1crega"  outgoing_chat_message   37553387   
11523  "01jb5a264ng5c42nv5ty3zr69g"  outgoing_chat_message   37553407   
11515  "01jb5a53hqx0thmb437emhzkx5"  outgoing_chat_message   37553407   
11522  "01jb5a3088t4668cmmse342860"  outgoing_chat_message   37553521   
11525  "01jb5a21ep1pnctz8bxeb1b76s"  outgoing_chat_message   37553613   
11508  "01jb5a7017mzwes4ar3yneghba"  outgoing_chat_message   37553625   
11600  "01jb5a8q4rvvyve87zq6e04jwy"  outgoing_chat_message   37553625   
11641  "01jb5ddeq21dy5np3zrrm9ztxg"  outgoing_chat_message   37553655   
11519  "01jb5a38fp8hamy1n30mtz47ch"  outgoing_chat_message   37553665   
6733   "01jb68a69jf4fe1e8q3vpe1x32"  outgoing_chat_message   37553677   
8370   "01jb68wfb45n0bnatv8f6zaj2n"  outgoing_chat_message   37553983   

       created_by                created_at             question_time  
10547     7996770 2024-10-27 09:25:39+03:00 2024-10-27 00:46:13+03:00  
11512     6744792 2024-10-27 00:02:35+03:00 2024-10-26 23:53:12+03:00  
1538      6645315 2024-10-26 09:14:39+03:00 2024-10-26 09:03:12+03:00  
11645    11490278 2024-10-27 01:02:42+03:00 2024-10-26 23:56:31+03:00  
868      10262497 2024-10-26 00:03:01+03:00 2024-10-26 00:02:07+03:00  
11599    11490270 2024-10-27 00:03:50+03:00 2024-10-26 23:56:34+03:00  
10545     7996770 2024-10-27 09:26:47+03:00 2024-10-27 06:09:25+03:00  
1562      7996770 2024-10-26 09:29:08+03:00 2024-10-26 01:18:14+03:00  
11510    11490270 2024-10-27 00:02:41+03:00 2024-10-26 23:55:14+03:00  
11514    11490270 2024-10-27 00:02:04+03:00 2024-10-26 23:49:36+03:00  
11596    11490270 2024-10-27 00:05:21+03:00 2024-10-26 23:57:12+03:00  
10544    11448610 2024-10-27 09:27:13+03:00 2024-10-27 07:44:33+03:00  
11518    11394130 2024

Тут встречаются и случаи, когда менеджер пришел на работу раньше и ответил на сообщение, когда менеджер задержался на работе и ответил после полуночи, а есть случаи, когда менеджер отвечал ночью. Разницы во времени ответа во всех этих случаях сильно разнятся, поэтому, так как в целом расчет среднего всегда чувствителен к аномальным значениям, я решил все эти случаи убрать из выборки.

Также мы убираем все входящие сообщения и исходящие, для которых не было вопроса от клиента(сделка состоит из одного исходящего сообщения)

In [386]:
df = df[(df['question_time'].notna()) & (df['type'] == 'outgoing_chat_message') & (~df['created_at'].dt.time.between(pd.Timestamp('00:00:00').time(), pd.Timestamp('09:30:00').time()))]

Далее определим функцию расчета разницы между временем вопроса и временем ответа.

1) Определем дату и время вопроса(q_time) и ответа(a_time);
2) Определяем переменную work_time_start, которяа будет определять начало рабочего дня и будем изменять дату в соответствии с датой ответа;
3) Для случая, когда вопрос был получен с полуночи до начала рабочего дня, разница будет определяться между временем ответа и временем начала рабочего дня;
4) Для случаев, когда вопрос был получен до полуночи, а ответ был отправлен утром после начала рабочего дня, разница будет считаться между временем ответа и вопроса за исключением 9 часов 30 минут, когда менеджер не находился на рабочем месте.
5) Во всех остальных случаях мы просто получаем разницу между временем ответа и временем вопроса

In [387]:
def calculate_time_diff(row):
    q_time = row['question_time']
    a_time = row['created_at']

    work_time_start = a_time.normalize() + pd.Timedelta(hours=9, minutes=30)

    if (q_time >= a_time.normalize()) and (q_time < work_time_start) and (a_time >= work_time_start):
        return a_time - work_time_start
    elif q_time < a_time.normalize() and a_time >= work_time_start:
        return a_time - q_time - pd.Timedelta(hours=9, minutes=30)
    else:
        return a_time - q_time

In [388]:
df['time_diff'] = df.apply(calculate_time_diff, axis=1)
df.head(20)

message_id                   type  entity_id  \
4102   "01jb4pncn7gm1re1zdxkzhnp6s"  outgoing_chat_message   35810509   
6639   "01jb4qj2skc2wjfdm8qaf69yjh"  outgoing_chat_message   35810509   
17852  "01jb7h4ad8weqjx1ynrh7s8k25"  outgoing_chat_message   35945535   
14833  "01jb72ezmaacke29h682geasez"  outgoing_chat_message   36099337   
15039  "01jb72nzkrzhcfhg9hfpbjt0vx"  outgoing_chat_message   36099337   
9238   "01jb4tsxdn395tp1jwknjajpwk"  outgoing_chat_message   36333761   
9569   "01jb4vbfpfwe4my2w5zhadeash"  outgoing_chat_message   36333761   
9547   "01jb4ve204kyab9ez1hb5d2e1f"  outgoing_chat_message   36333761   
9801   "01jb4vhn2hpdj09h1wre5ctzkn"  outgoing_chat_message   36333761   
9786   "01jb4vkdrhrw67wsn7gxxdvt02"  outgoing_chat_message   36333761   
6117   "01jb4w4gbgyf1mwe93fvhaf9wc"  outgoing_chat_message   36333761   
9909   "01jb4zbhe225cdxsz4nsava9g3"  outgoing_chat_message   36333761   
10078  "01jb4zs5pk371qh9xj6r7cb4ta"  outgoing_chat_message   36333761   
10216  "01jb501hjdfg05ps9pxcfm6vf1"  outgoing_chat_message   36333761   
10210  "01jb5042g5f0p9cerrj1mxznbr"  outgoing_chat_message   36333761   
11031  "01jb58p87bpndpn0w3rxjpapzv"  outgoing_chat_message   36333761   
10979  "01jb6b3kzd52p7t0h5bypv4k06"  outgoing_chat_message   36373713   
10962  "01jb6b4zmkxzfsrvp2q5n2r3mj"  outgoing_chat_message   36373713   
10904  "01jb6b7ykv1a9dhscb99ttkn90"  outgoing_chat_message   36373713   
10885  "01jb6bamgt9bnd3089cqjdcs4c"  outgoing_chat_message   36373713   

       created_by                created_at             question_time  \
4102     10262493 2024-10-26 18:21:05+03:00 2024-10-26 18:19:26+03:00   
6639     10262493 2024-10-26 18:36:46+03:00 2024-10-26 18:34:46+03:00   
17852    10262493 2024-10-27 20:42:07+03:00 2024-10-27 20:38:36+03:00   
14833    11241942 2024-10-27 16:25:47+03:00 2024-10-27 16:21:56+03:00   
15039    11241942 2024-10-27 16:29:37+03:00 2024-10-27 16:27:14+03:00   
9238     10262485 2024-10-26 19:33:28+03:00 2024-10-26 19:28:45+03:00   
9569     10262485 2024-10-26 19:43:04+03:00 2024-10-26 19:37:02+03:00   
9547     10262485 2024-10-26 19:44:28+03:00 2024-10-26 19:43:27+03:00   
9801     10262485 2024-10-26 19:46:26+03:00 2024-10-26 19:45:15+03:00   
9786     10262485 2024-10-26 19:47:24+03:00 2024-10-26 19:46:57+03:00   
6117     10262485 2024-10-26 19:56:44+03:00 2024-10-26 19:47:46+03:00   
9909     10262485 2024-10-26 20:53:00+03:00 2024-10-26 20:50:46+03:00   
10078    10262485 2024-10-26 21:00:27+03:00 2024-10-26 20:53:53+03:00   
10216    10262485 2024-10-26 21:05:01+03:00 2024-10-26 21:04:43+03:00   
10210    10262485 2024-10-26 21:06:24+03:00 2024-10-26 21:05:41+03:00   
11031    10262485 2024-10-26 23:36:08+03:00 2024-10-26 23:35:18+03:00   
10979     7996770 2024-10-27 09:37:38+03:00 2024-10-27 09:35:47+03:00   
10962     7996770 2024-10-27 09:38:22+03:00 2024-10-27 09:37:55+03:00   
10904     7996770 2024-10-27 09:40:00+03:00 2024-10-27 09:39:30+03:00   
10885     7996770 2024-10-27 09:41:28+03:00 2024-10-27 09:40:32+03:00   

            time_diff  
4102  0 days 00:01:39  
6639  0 days 00:02:00  
17852 0 days 00:03:31  
14833 0 days 00:03:51  
15039 0 days 00:02:23  
9238  0 days 00:04:43  
9569  0 days 00:06:02  
9547  0 days 00:01:01  
9801  0 days 00:01:11  
9786  0 days 00:00:27  
6117  0 days 00:08:58  
9909  0 days 00:02:14  
10078 0 days 00:06:34  
10216 0 days 00:00:18  
10210 0 days 00:00:43  
11031 0 days 00:00:50  
10979 0 days 00:01:51  
10962 0 days 00:00:27  
10904 0 days 00:00:30  
10885 0 days 00:00:56

В качестве примеров корректной работы функции посмотрим несколько случаев

In [389]:
df[df['entity_id'] == 36698385].iloc[[0], :]

message_id                   type  entity_id  \
2233  "01jb3xb2ytv0w3fpnnvmm8gfmh"  outgoing_chat_message   36698385   

      created_by                created_at             question_time  \
2233    10262485 2024-10-26 10:58:33+03:00 2024-10-26 08:19:55+03:00   

           time_diff  
2233 0 days 01:28:33

Здесь мы видим, что вопрос был получен в 08:19 утра, а ответ отправлен в 10:58. В таком случае разница должна считается от момента начала рабочего дня, и равняться 01:28:00. Так и есть.

Другой пример:

In [390]:
df[df['entity_id'] == 36827109].iloc[[0], :]

message_id                   type  entity_id  \
11295  "01jb6bykhsdsc9e1q3fma0km70"  outgoing_chat_message   36827109   

       created_by                created_at             question_time  \
11295    11188802 2024-10-27 09:52:22+03:00 2024-10-26 23:22:32+03:00   

            time_diff  
11295 0 days 00:59:50

Здесь вопрос был получен до завершения рабочего дня(в 23:22:32), а ответ отправлен в 09:52:22 утром следующего дня. Как видим, функция здесь также отрабатывает корректно и вместо 10 с половиной часов разницы мы получаем разницу около часа.

Теперь переведем полученное значения разницы в минуты для лучшего восприятия. Для этого сначала переведем тип в формат секунд, возьмем их целое количесвто и разделим на 60

In [391]:
df['time_diff_min'] = round(df['time_diff'].astype('timedelta64[s]').astype('int64') / 60, 2)

df.head(20)

message_id                   type  entity_id  \
4102   "01jb4pncn7gm1re1zdxkzhnp6s"  outgoing_chat_message   35810509   
6639   "01jb4qj2skc2wjfdm8qaf69yjh"  outgoing_chat_message   35810509   
17852  "01jb7h4ad8weqjx1ynrh7s8k25"  outgoing_chat_message   35945535   
14833  "01jb72ezmaacke29h682geasez"  outgoing_chat_message   36099337   
15039  "01jb72nzkrzhcfhg9hfpbjt0vx"  outgoing_chat_message   36099337   
9238   "01jb4tsxdn395tp1jwknjajpwk"  outgoing_chat_message   36333761   
9569   "01jb4vbfpfwe4my2w5zhadeash"  outgoing_chat_message   36333761   
9547   "01jb4ve204kyab9ez1hb5d2e1f"  outgoing_chat_message   36333761   
9801   "01jb4vhn2hpdj09h1wre5ctzkn"  outgoing_chat_message   36333761   
9786   "01jb4vkdrhrw67wsn7gxxdvt02"  outgoing_chat_message   36333761   
6117   "01jb4w4gbgyf1mwe93fvhaf9wc"  outgoing_chat_message   36333761   
9909   "01jb4zbhe225cdxsz4nsava9g3"  outgoing_chat_message   36333761   
10078  "01jb4zs5pk371qh9xj6r7cb4ta"  outgoing_chat_message   36333761   
10216  "01jb501hjdfg05ps9pxcfm6vf1"  outgoing_chat_message   36333761   
10210  "01jb5042g5f0p9cerrj1mxznbr"  outgoing_chat_message   36333761   
11031  "01jb58p87bpndpn0w3rxjpapzv"  outgoing_chat_message   36333761   
10979  "01jb6b3kzd52p7t0h5bypv4k06"  outgoing_chat_message   36373713   
10962  "01jb6b4zmkxzfsrvp2q5n2r3mj"  outgoing_chat_message   36373713   
10904  "01jb6b7ykv1a9dhscb99ttkn90"  outgoing_chat_message   36373713   
10885  "01jb6bamgt9bnd3089cqjdcs4c"  outgoing_chat_message   36373713   

       created_by                created_at             question_time  \
4102     10262493 2024-10-26 18:21:05+03:00 2024-10-26 18:19:26+03:00   
6639     10262493 2024-10-26 18:36:46+03:00 2024-10-26 18:34:46+03:00   
17852    10262493 2024-10-27 20:42:07+03:00 2024-10-27 20:38:36+03:00   
14833    11241942 2024-10-27 16:25:47+03:00 2024-10-27 16:21:56+03:00   
15039    11241942 2024-10-27 16:29:37+03:00 2024-10-27 16:27:14+03:00   
9238     10262485 2024-10-26 19:33:28+03:00 2024-10-26 19:28:45+03:00   
9569     10262485 2024-10-26 19:43:04+03:00 2024-10-26 19:37:02+03:00   
9547     10262485 2024-10-26 19:44:28+03:00 2024-10-26 19:43:27+03:00   
9801     10262485 2024-10-26 19:46:26+03:00 2024-10-26 19:45:15+03:00   
9786     10262485 2024-10-26 19:47:24+03:00 2024-10-26 19:46:57+03:00   
6117     10262485 2024-10-26 19:56:44+03:00 2024-10-26 19:47:46+03:00   
9909     10262485 2024-10-26 20:53:00+03:00 2024-10-26 20:50:46+03:00   
10078    10262485 2024-10-26 21:00:27+03:00 2024-10-26 20:53:53+03:00   
10216    10262485 2024-10-26 21:05:01+03:00 2024-10-26 21:04:43+03:00   
10210    10262485 2024-10-26 21:06:24+03:00 2024-10-26 21:05:41+03:00   
11031    10262485 2024-10-26 23:36:08+03:00 2024-10-26 23:35:18+03:00   
10979     7996770 2024-10-27 09:37:38+03:00 2024-10-27 09:35:47+03:00   
10962     7996770 2024-10-27 09:38:22+03:00 2024-10-27 09:37:55+03:00   
10904     7996770 2024-10-27 09:40:00+03:00 2024-10-27 09:39:30+03:00   
10885     7996770 2024-10-27 09:41:28+03:00 2024-10-27 09:40:32+03:00   

            time_diff  time_diff_min  
4102  0 days 00:01:39           1.65  
6639  0 days 00:02:00           2.00  
17852 0 days 00:03:31           3.52  
14833 0 days 00:03:51           3.85  
15039 0 days 00:02:23           2.38  
9238  0 days 00:04:43           4.72  
9569  0 days 00:06:02           6.03  
9547  0 days 00:01:01           1.02  
9801  0 days 00:01:11           1.18  
9786  0 days 00:00:27           0.45  
6117  0 days 00:08:58           8.97  
9909  0 days 00:02:14           2.23  
10078 0 days 00:06:34           6.57  
10216 0 days 00:00:18           0.30  
10210 0 days 00:00:43           0.72  
11031 0 days 00:00:50           0.83  
10979 0 days 00:01:51           1.85  
10962 0 days 00:00:27           0.45  
10904 0 days 00:00:30           0.50  
10885 0 days 00:00:56           0.93

Теперь мы можем рассчитать среднее количество минут на ответ каждого менеджера. Группировку производим по столбцу created_by.

In [392]:
df['avg_time_diff_min'] = round(df.groupby('created_by')['time_diff_min'].transform('mean'), 1)
df.head(20)

message_id                   type  entity_id  \
4102   "01jb4pncn7gm1re1zdxkzhnp6s"  outgoing_chat_message   35810509   
6639   "01jb4qj2skc2wjfdm8qaf69yjh"  outgoing_chat_message   35810509   
17852  "01jb7h4ad8weqjx1ynrh7s8k25"  outgoing_chat_message   35945535   
14833  "01jb72ezmaacke29h682geasez"  outgoing_chat_message   36099337   
15039  "01jb72nzkrzhcfhg9hfpbjt0vx"  outgoing_chat_message   36099337   
9238   "01jb4tsxdn395tp1jwknjajpwk"  outgoing_chat_message   36333761   
9569   "01jb4vbfpfwe4my2w5zhadeash"  outgoing_chat_message   36333761   
9547   "01jb4ve204kyab9ez1hb5d2e1f"  outgoing_chat_message   36333761   
9801   "01jb4vhn2hpdj09h1wre5ctzkn"  outgoing_chat_message   36333761   
9786   "01jb4vkdrhrw67wsn7gxxdvt02"  outgoing_chat_message   36333761   
6117   "01jb4w4gbgyf1mwe93fvhaf9wc"  outgoing_chat_message   36333761   
9909   "01jb4zbhe225cdxsz4nsava9g3"  outgoing_chat_message   36333761   
10078  "01jb4zs5pk371qh9xj6r7cb4ta"  outgoing_chat_message   36333761   
10216  "01jb501hjdfg05ps9pxcfm6vf1"  outgoing_chat_message   36333761   
10210  "01jb5042g5f0p9cerrj1mxznbr"  outgoing_chat_message   36333761   
11031  "01jb58p87bpndpn0w3rxjpapzv"  outgoing_chat_message   36333761   
10979  "01jb6b3kzd52p7t0h5bypv4k06"  outgoing_chat_message   36373713   
10962  "01jb6b4zmkxzfsrvp2q5n2r3mj"  outgoing_chat_message   36373713   
10904  "01jb6b7ykv1a9dhscb99ttkn90"  outgoing_chat_message   36373713   
10885  "01jb6bamgt9bnd3089cqjdcs4c"  outgoing_chat_message   36373713   

       created_by                created_at             question_time  \
4102     10262493 2024-10-26 18:21:05+03:00 2024-10-26 18:19:26+03:00   
6639     10262493 2024-10-26 18:36:46+03:00 2024-10-26 18:34:46+03:00   
17852    10262493 2024-10-27 20:42:07+03:00 2024-10-27 20:38:36+03:00   
14833    11241942 2024-10-27 16:25:47+03:00 2024-10-27 16:21:56+03:00   
15039    11241942 2024-10-27 16:29:37+03:00 2024-10-27 16:27:14+03:00   
9238     10262485 2024-10-26 19:33:28+03:00 2024-10-26 19:28:45+03:00   
9569     10262485 2024-10-26 19:43:04+03:00 2024-10-26 19:37:02+03:00   
9547     10262485 2024-10-26 19:44:28+03:00 2024-10-26 19:43:27+03:00   
9801     10262485 2024-10-26 19:46:26+03:00 2024-10-26 19:45:15+03:00   
9786     10262485 2024-10-26 19:47:24+03:00 2024-10-26 19:46:57+03:00   
6117     10262485 2024-10-26 19:56:44+03:00 2024-10-26 19:47:46+03:00   
9909     10262485 2024-10-26 20:53:00+03:00 2024-10-26 20:50:46+03:00   
10078    10262485 2024-10-26 21:00:27+03:00 2024-10-26 20:53:53+03:00   
10216    10262485 2024-10-26 21:05:01+03:00 2024-10-26 21:04:43+03:00   
10210    10262485 2024-10-26 21:06:24+03:00 2024-10-26 21:05:41+03:00   
11031    10262485 2024-10-26 23:36:08+03:00 2024-10-26 23:35:18+03:00   
10979     7996770 2024-10-27 09:37:38+03:00 2024-10-27 09:35:47+03:00   
10962     7996770 2024-10-27 09:38:22+03:00 2024-10-27 09:37:55+03:00   
10904     7996770 2024-10-27 09:40:00+03:00 2024-10-27 09:39:30+03:00   
10885     7996770 2024-10-27 09:41:28+03:00 2024-10-27 09:40:32+03:00   

            time_diff  time_diff_min  avg_time_diff_min  
4102  0 days 00:01:39           1.65                7.8  
6639  0 days 00:02:00           2.00                7.8  
17852 0 days 00:03:31           3.52                7.8  
14833 0 days 00:03:51           3.85                7.6  
15039 0 days 00:02:23           2.38                7.6  
9238  0 days 00:04:43           4.72               15.0  
9569  0 days 00:06:02           6.03               15.0  
9547  0 days 00:01:01           1.02               15.0  
9801  0 days 00:01:11           1.18               15.0  
9786  0 days 00:00:27           0.45               15.0  
6117  0 days 00:08:58           8.97               15.0  
9909  0 days 00:02:14           2.23               15.0  
10078 0 days 00:06:34           6.57               15.0  
10216 0 days 00:00:18           0.30               15.0  
10210 0 days 00:00:43           0.72               15.0  
11031 0 days 00:00:50        

Теперь скомпануем итоговую сводную таблицу. Сначала соединим текущую таблицу с датасетом, содержащим информацию о менеджерах. Выберем только нужные столбцы и уберем дубликаты.

In [393]:
df_final = pd.merge(df, df_managers, left_on='created_by', right_on='mop_id', how='right')[['mop_id', 'name_mop', 'rop_id', 'avg_time_diff_min']].drop_duplicates()
df_final

mop_id             name_mop  rop_id  avg_time_diff_min
0      6645315       Гюнель и Илина       1                8.6
1044   6744792         Юля и Наташа       1                9.5
1890  10262505         Вика и Марго       1                NaN
1891  10262513           Ира и Варя       1               20.0
2367  10465254         Настя и Даша       1                3.5
2691  11448610          Вика и Катя       1               12.9
3770  11490270          Аня и Ксюша       1               13.8
4410  11490278      Ксюша и Джамиля       1               12.3
4682   6780177          Даша и Даша       2                4.0
4832   7436430         Софа и Алина       2                NaN
4833   7996770           Лиза и Ева       2                6.0
5519  10262485          Соня и Катя       2               15.0
5788  10262497         Мария и Соня       2                5.3
5982  11188802         Даша и Влада       2               14.2
6259  11394130         Настя и Саша       2                5.8
7043   7408305        Влада и Настя       3                4.7
7089   7408332             Пустой 2       3                NaN
7090   7417617     Порхачева Полина       3                8.4
7104   8558094          Алина и Юля       3                4.6
7334  10262493          Ками и Мила       3                7.8
7559  10465690       Настя и Малика       3                5.6
7807  11241942        Даша и Карина       3                7.6
7952  11394126       Полина Мирзоян       3               10.4
8000  11394134         Софья Боднар       3                8.7
8101  11503694     Ахмедова Джамиля       3                NaN
8102  11550890  Ангелина Милованова       3                6.1

Теперь добавим в финальную сводную таблицу имя руководителя каждого менеджера. Снова используем метод merge. В конце отсортируемся по среднему значению ответа

In [394]:
df_final = pd.merge(df_final, df_rops, on='rop_id', how='inner')[['mop_id', 'name_mop', 'rop_name', 'avg_time_diff_min']].sort_values('avg_time_diff_min')
df_final

mop_id             name_mop    rop_name  avg_time_diff_min
4   10465254         Настя и Даша    Катя РОП                3.5
8    6780177          Даша и Даша  Полина РОП                4.0
18   8558094          Алина и Юля     Эля РОП                4.6
15   7408305        Влада и Настя     Эля РОП                4.7
12  10262497         Мария и Соня  Полина РОП                5.3
20  10465690       Настя и Малика     Эля РОП                5.6
14  11394130         Настя и Саша  Полина РОП                5.8
10   7996770           Лиза и Ева  Полина РОП                6.0
25  11550890  Ангелина Милованова     Эля РОП                6.1
21  11241942        Даша и Карина     Эля РОП                7.6
19  10262493          Ками и Мила     Эля РОП                7.8
17   7417617     Порхачева Полина     Эля РОП                8.4
0    6645315       Гюнель и Илина    Катя РОП                8.6
23  11394134         Софья Боднар     Эля РОП                8.7
1    6744792         Юля и Наташа    Катя РОП                9.5
22  11394126       Полина Мирзоян     Эля РОП               10.4
7   11490278      Ксюша и Джамиля    Катя РОП               12.3
5   11448610          Вика и Катя    Катя РОП               12.9
6   11490270          Аня и Ксюша    Катя РОП               13.8
13  11188802         Даша и Влада  Полина РОП               14.2
11  10262485          Соня и Катя  Полина РОП               15.0
3   10262513           Ира и Варя    Катя РОП               20.0
2   10262505         Вика и Марго    Катя РОП                NaN
9    7436430         Софа и Алина  Полина РОП                NaN
16   7408332             Пустой 2     Эля РОП                NaN
24  11503694     Ахмедова Джамиля     Эля РОП                NaN